### PR TITLE
Unet++

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
- Add UNet++ model

- Add multiple outputs to body

### CODE REVIEW

File: batchflow/models/tf/unet.py [hunk 1]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+
+        x = super().decoder(inputs, name=name, filters=decoder_filters,
+                            return_all=False, **{**kwargs, 'num_stages': num_stages})
+
+        semantic_outputs.append(x)
+
+        return [semantic_outputs[i] for i in n_outputs]
+
+    @classmethod
+    def head(cls, inputs, targets, name='head', **kwargs):
```
Comment (roman-kh) [inline batchflow/models/tf/unet.py]: Document somewhere that `head` always expects a sequence of inputs

File: batchflow/models/tf/unet.py [hunk 2]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+
+        semantic_outputs = []
+        n_outputs = kwargs['n_outputs']
+        if isinstance(n_outputs, int):
+            n_outputs = list(range(-n_outputs, 0))
+
+        for i in range(1, num_stages):
+            _inputs = inputs[:i] + [inputs[i]] * 2
+
+            _kwargs = kwargs
```
Comment (roman-kh) [inline batchflow/models/tf/unet.py]: Not needed

File: batchflow/models/tf/unet.py [hunk 3]
```diff
@@ -58,14 +66,7 @@ class UNet(EncoderDecoder):
...
 
-        config['body/encoder/num_stages'] = 4
-        config['body/encoder/order'] = ['block', 'skip', 'downsampling']
-        config['body/encoder/blocks'] += dict(layout='cna cna', kernel_size=3, filters=[64, 128, 256, 512])
-        config['body/embedding'] += dict(layout='cna cna', kernel_size=3, filters=1024)
-        config['body/decoder/order'] = ['upsampling', 'combine', 'block']
-        config['body/decoder/blocks'] += dict(layout='cna cna', kernel_size=3, filters=[512, 256, 128, 64])
-
-        config['loss'] = 'ce'
+        config['body/n_outputs'] = 1
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: Why did you delete most of the default config?

File: batchflow/models/tf/base.py [hunk 4]
```diff
@@ -1587,13 +1587,13 @@ def head(cls, inputs, name='head', **kwargs):
             return conv_block(inputs, name=name, **kwargs)
         return inputs
 
-    def output(self, inputs, predictions=None, ops=None, prefix=None, **kwargs):
+    def output(self, inputs, predictions=None, ops=None, **kwargs):
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/base.py]: Can you elaborate on what is going on here besides some code refactoring (better name making for added output `ops` and removing redundant `prefix` argument)?

File: batchflow/models/tf/unet.py [hunk 5]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+            config['body/decoder/dense/kernel'] = 3
+
+        if config.get('body/decoder/dense/combine_op') is None:
+            config['body/decoder/dense/combine_op'] = 'concat'
+
+        return config
+
+    @classmethod
+    def decoder(cls, inputs, name='decoder', **kwargs):
+        decoder_filters = kwargs['upsample']['filters']
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]:
```suggestion
        decoder_filters = kwargs['upsample/filters']
```
can improve readability everywhere

File: batchflow/models/tf/unet.py [hunk 6]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+                        'data_format': kwargs.get('data_format'),
+                        **(combine_op if isinstance(combine_op, dict) else {})}
+
+        semantic_outputs = []
+        n_outputs = kwargs['n_outputs']
+        if isinstance(n_outputs, int):
+            n_outputs = list(range(-n_outputs, 0))
+
+        for i in range(1, num_stages):
+            _inputs = inputs[:i] + [inputs[i]] * 2
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: `_inputs` is used only once, so you don't really need it here

File: batchflow/models/tf/unet.py [hunk 7]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+
+            _kwargs = kwargs
+            _kwargs['upsample'] = _kwargs['dense']['upsample']
+            _kwargs['upsample']['filters'] = decoder_filters[-i:]
+            _kwargs['blocks']['filters'] = decoder_filters[-i:]
+
+            outputs = super().decoder(_inputs, name=name,
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: Same `name` would create scopes like `decoder`, `decoder_1`, `decoder_2` etc, while it can be better to reserve name `decoder` for the deepest branch

File: batchflow/models/tf/unet.py [hunk 8]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+
+                dense_args = {**unpack_args(kwargs['dense'], j, num_stages), **kwargs}
+                x = conv_block(x, name='x-{}-{}'.format(j, i), **dense_args)
+
+                if j == 0:
+                    semantic_outputs.append(x)
+
+                inputs[j] = combine([x, inputs[j]], name='concat-{}-{}'.format(j, i-j), **combine_args)
+
+        kwargs['upsample']['filters'] = decoder_filters
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: Do you really need it? You don't seem to alter `kwargs` in any way

File: batchflow/models/tf/unet.py (hunk 2 above)
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: Make a proper copy

File: batchflow/models/tf/unet.py [hunk 9]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+            _kwargs = kwargs
+            _kwargs['upsample'] = _kwargs['dense']['upsample']
+            _kwargs['upsample']['filters'] = decoder_filters[-i:]
+            _kwargs['blocks']['filters'] = decoder_filters[-i:]
+
+            outputs = super().decoder(_inputs, name=name,
+                                      return_all=True, **{**_kwargs, 'num_stages': i})
+
+            for j, x in enumerate(inputs[:i]):
+                x = combine([x, outputs[::-1][j]], **combine_args)
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]:
```suggestion
                x = combine([x, outputs[-j]], **combine_args)
```
that would be easier to read. Or you could use `zip` over `reversed` as iterator

File: batchflow/models/tf/unet.py [hunk 10]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+
+                dense_args = {**unpack_args(kwargs['dense'], j, num_stages), **kwargs}
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]:
```suggestion
                dense_args = {**kwargs, **unpack_args(kwargs['dense'], j, num_stages)}
```
Current version seems to be working only because `conv_block` applies essentially the same `unpack_args` under the hood. You can even remove this `unpacking`, but explicit is better that implicit, right?

File: batchflow/models/tf/unet.py [hunk 11]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+                x = conv_block(x, name='x-{}-{}'.format(j, i), **dense_args)
+
+                if j == 0:
+                    semantic_outputs.append(x)
+
+                inputs[j] = combine([x, inputs[j]], name='concat-{}-{}'.format(j, i-j), **combine_args)
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: Can't find that in the original paper, though it does not seem to be a bad idea

File: batchflow/models/tf/unet.py [hunk 12]
```diff
@@ -87,3 +88,140 @@ def build_config(self, names=None):
...
+        semantic_outputs.append(x)
+
+        return [semantic_outputs[i] for i in n_outputs]
+
+    @classmethod
+    def head(cls, inputs, targets, name='head', **kwargs):
+        res = []
+        for i, x in enumerate(inputs):
+            res.append(super().head(x, targets, name=name+'-'+str(i), **kwargs))
+        return res
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/unet.py]: What will happen to this list afterwards?

File: batchflow/models/tf/base.py [hunk 13]
```diff
@@ -1648,74 +1648,100 @@ def output(self, inputs, predictions=None, ops=None, prefix=None, **kwargs):
...
-    def _add_output_callable(self, inputs, oper, name, attr_prefix, device, **kwargs):
+    def _add_output_callable(self, inputs, oper, name, attr_name, device, **kwargs):
         _ = kwargs
         x = oper(inputs)
         name = name or oper.__name__
-        self.store_to_attr(attr_prefix + name, x, device)
+        self.store_to_attr(attr_name, x, device)
         return x
 
+    def _add_predictions(self, predictions, name, attr_name, **kwargs):
```
Comment (SergeyTsimfer) [inline batchflow/models/tf/base.py]: It seems strange to always `stack` `head` outputs: what if you just wanted to have multiple outputs to process them individually in your `loss` function? That is, for example, the case for seismic noise attenuation models

File: batchflow/models/utils.py [hunk 14]
```diff
@@ -5,7 +5,7 @@ def unpack_args(args, layer_no, layers_max):
     """ Return layer parameters """
     new_args = {}
     for arg in args:
-        if isinstance(args[arg], list) and layers_max > 1:
+        if isinstance(args[arg], list):
```
[the remark this answers is not in the document]
Comment (AlexeyKozhevin) [reply batchflow/models/utils.py]: OK, I will run them

File: batchflow/models/tf/unet.py (hunk 3 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: Thanks, I've edited wrong class.

File: batchflow/models/tf/unet.py (hunk 5 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: `kwargs` is ordinal dict

File: batchflow/models/tf/base.py (hunk 13 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/base.py]: You can use outputs separately by names `predictions_0` `predictions_1` etc.

File: batchflow/models/tf/unet.py (hunk 6 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: I think that in that case it is better to define variable then write
```
            outputs = super().decoder(inputs[:i] + [inputs[i]] * 2, name=name,
                                      return_all=True, **{**kwargs, 'num_stages': I})
```

File: batchflow/models/tf/unet.py (hunk 11 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: Authors proposed to implement dense blocks, I've realised it in that way.

File: batchflow/models/tf/unet.py (hunk 12 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: It will be used to create outputs

File: batchflow/models/tf/base.py (hunk 4 above)
Comment (AlexeyKozhevin) [reply batchflow/models/tf/base.py]: Previously, output-block processed multiple outputs from `head` in wrong way.
Now all outputs from `head` will be transformed by `ops` dict and in case of multiple outputs each output will have postfix `_{i}` where i is an index.

File: batchflow/models/tf/unet.py (hunk 9 above)
Comment (roman-kh) [reply batchflow/models/tf/unet.py]: Do not forget about `0`: `[-j-1]`

Comment (AlexeyKozhevin) [reply batchflow/models/tf/unet.py]: I've rewritten it in last commit as 
`x = combine([x, outputs[-j-1]], **combine_args)`
